### PR TITLE
[MLIR][OpenMP] Clause-based OpenMP operation definition

### PR DESCRIPTION
**Pull request by skatrak**
This patch updates `OpenMP_Op` definitions to be based on the new set of `OpenMP_Clause` definitions, and to take advantage of clause-based automatically-generated argument lists, descriptions, assembly format and class declarations.

There are also changes introduced to the clause operands structures to match the current set of tablegen clause definitions. These two are very closely linked and should be kept in sync. It would probably be a good idea to try generating clause operands structures from the tablegen `OpenMP_Clause` definitions in the future.

As a result of this change, arguments for some operations have been reordered. This patch also addresses this by updating affected operation build calls and unit tests. Some other updates to tests related to the order of arguments in the resulting assembly format and others due to certain previous inconsistencies in the printing/parsing of clauses are addressed.

The printer and parser functions for the `map` clause are updated, so that they are able to handle `map` clauses linked to entry block arguments as well as those which aren't.

### CODE REVIEW

File: mlir/include/mlir/Dialect/OpenMP/OpenMPOps.td
Comment (tblah) [inline mlir/include/mlir/Dialect/OpenMP/OpenMPOps.td:270]: Shouldn't this use the machinery in `OpenMP_ReductionClause`?

File: mlir/include/mlir/Dialect/OpenMP/OpenMPOps.td
Comment (tblah) [inline mlir/include/mlir/Dialect/OpenMP/OpenMPOps.td:668]: Here too. I guess we need separate work to standardize how reductions work over different ops?

**Comment by skatrak** [comment]
@tblah thanks for giving this a look. Basically the issue here with the reduction clause is that its description is currently different for each operation that accepts it. They are generally quite similar, just not the same. I think the clause is already consistent in terms of how it's represented but what it means is slightly different depending on the operation, hence the variations in their way of describing it. If you look at the `OpenMP_ReductionClause` definition in OpenMPClauses.td, you'll see that I added a comment to it saying basically that no description is defined for it because it's different on each operation (same thing I did for the `if` and `in_reduction` clauses).

It would be great to agree on a generic description for that clause that would work for all ops, so that it makes sense while reading the generated documentation while being able to centralize this into the clause definition rather than each operation. I created a gist [here](https://gist.github.com/skatrak/ee6e1829980867f8119f91f84cb284b1) with the whole list of descriptions that currently exist for `reduction`, in case that helps with figuring out a shared description. Another option would be to provide a generic description for that clause used by most operations and then override it for one or two if they need to be more specific about something.

**Comment by tblah** [comment]
> @tblah thanks for giving this a look. Basically the issue here with the reduction clause is that its description is currently different for each operation that accepts it. 

> It would be great to agree on a generic description for that clause that would work for all ops, so that it makes sense while reading the generated documentation and also being centralized into the clause definition rather than each operation. I created a gist [here](https://gist.github.com/skatrak/ee6e1829980867f8119f91f84cb284b1) with the whole list of descriptions that currently exist for `reduction`, in case that helps with figuring out a shared description. 

Thanks for taking the time to create the gist. Some thoughts
- I think some of the descriptions are out of date. `omp.reduction` has been removed: https://github.com/llvm/llvm-project/pull/92732
- Assuming that `omp.reduction` really was unused, I think that leaves byref as the main difference?

I guess fixing byref is on me (https://github.com/llvm/llvm-project/pull/92244). Unfortunately I can't work on this immediately so I won't hold up this PR for it.

**Review by tblah — APPROVED** [review]
LGTM, but please wait for another reviewer

Edit: and CI is red

**Comment by skatrak** [comment]
Quick update about reduction clause descriptions @tblah: I updated the previous PR on the stack (#92521) to add a shared description that basically incorporates all of the information that was spread across the various operations using it. I updated the description of `omp.taskloop` in this PR as well to keep the additional related information it previously had while still inheriting the base description for the clause.

Also, it's expected that this PR breaks buildbots because of some small reordering of operation arguments and assembly format normalization. The next PR in the stack (#92524) addresses this.

**Comment by skatrak** [comment]
> > I guess fixing byref is on me (#92244). Unfortunately I can't work on this immediately so I won't hold up this PR for it.
> 
> @skatrak does #96215 cover everything you need?

Thank you for the heads-up, that certainly helps. My plan is to update the PR stack after yours and one or two other PRs that conflict with this change land and hopefully by then all patches in this stack should be reviewed/approved to be merged.

**Review by TIFitis — APPROVED** [review]
I'm happy with the patch. Thanks for the good work :)

**Pull request merged**
